### PR TITLE
Remove redundant CPU decomposition when done also on GPU

### PR DESCRIPTION
**Pull request by razvnane**
Factor out the initialization code needed before doing a ILU decomposition to remove the redundant software decomposition when we also perform it on the GPU. This was previously done in #4558, but lost in the adaptation to ISTLSolverEbosBda.

### CODE REVIEW

File: opm/simulators/linalg/ISTLSolverEbosBda.hpp [hunk 1]
```diff
@@ -182,7 +182,15 @@ class ISTLSolverEbosBda : public ISTLSolverEbos<TypeTag>
     {
         OPM_TIMEBLOCK(prepare);
         [[maybe_unused]] const bool firstcall = (this->matrix_ == nullptr);
-        ParentType::prepare(M,b);
+
+        // Avoid performing the decomposition on CPU when we also do it on GPU, but we do
+        // need to initialize the pointers.
+         if(bdaBridge_){
```
Comment (atgeirr) [inline opm/simulators/linalg/ISTLSolverEbosBda.hpp:188]: Formatting and indentation.

**Review by atgeirr — APPROVED** [review]
Should be merged. Minor reformatting requested.

**Comment by blattms** [comment]
Not sure whether this code is still there/functional and my memory even reflects reality.
What happens for CPR on the GPU. Didn't that somehow use the AMG hierarchy from the CPU? With this it won't be there anymore.

**Comment by razvnane** [comment]
It is functional and it works similarly for CPR as well. I think there was a misunderstanding about what was used from CPU, but what is reused are the DUNE classes and functions regarding the creation of the CPR. These are called from the create_preconditioner_amg() function from opencl/CPR.cpp and do not depend on ISTLSolverEbos::prepare() function being called beforehand. I also tested it timing the first step of running NORNE and it shows what we would expect, i.e., a reduction in the solve time by removing the time for the preparation on the CPU (note: in the timing results the _prepare_ time is included in the _solve_ time):
```
<with removal>

CPR create_preconditioner bilu0(): 0.0608781 s
CPR create_preconditioner_amg(): 0.141993 s 
openclSolver::create_preconditioner(): 0.202939 s
openclSolver::copy_system_to_gpu(): 0.00479971 s 
openclSolver::prec_apply:  0.0482614 s
openclSolver::total_solve: 0.0539096 s
openclSolver::solve_system(): 0.0540114 s
openclSolver::get_result(): 0.000339837 s 
BlackoilModelEbos::solveJacobianSystem cum prepare time: 0.000239601
BlackoilModelEbos::solveJacobianSystem cum solve time: 0.27242
```
```
<without removal>

CPR create_preconditioner bilu0(): 0.0465751 s
CPR create_preconditioner_amg(): 0.133502 s
openclSolver::create_preconditioner(): 0.180129 s
openclSolver::copy_system_to_gpu(): 0.00479273 s
openclSolver::prec_apply:  0.0475754 s
openclSolver::total_solve: 0.0576826 s
openclSolver::solve_system(): 0.0577893 s
openclSolver::get_result(): 0.00031434 s
BlackoilModelEbos::solveJacobianSystem prepare time: 0.294071
BlackoilModelEbos::solveJacobianSystem solve time: 0.546205
```

**Comment by atgeirr** [comment]
jenkins build this please

**Pull request merged**